tenders/hvt/hvt_core.c

```diff
@@ -121,62 +136,169 @@ static void hypercall_walltime(struct hvt *hvt, hvt_gpa_t gpa)
 {
-    struct hvt_poll *t =
-        HVT_CHECKED_GPA_P(hvt, gpa, sizeof (struct hvt_poll));
+    struct hvt_hc_poll *t =
+        HVT_CHECKED_GPA_P(hvt, gpa, sizeof (struct hvt_hc_poll));
+
+#if defined(__linux__)
+    /*
+     * On Linux, in order to support nanosecond timeouts, as defined by the
+     * Solo5 API, we use a timerfd internally in the waitset. Account for this
+     * in the number of requested events.
+     */
+    int nevents = npollfds ? (npollfds + 1) : 1;
+    int nrevents;
+    uint64_t ready_set = 0;
+
+    struct epoll_event revents[nevents];
+    struct itimerspec it = {
+        .it_interval = { 0 },
+        .it_value = {
+            .tv_sec = t->timeout_nsecs / 1000000000ULL,
+            .tv_nsec = t->timeout_nsecs % 1000000000ULL
+        }
+    };
+    if (timerfd_settime(timerfd, 0, &it, NULL) == -1)
+        err(1, "timerfd_settime() failed");
+    /*
+     * We can always safely restart this call on EINTR, since the internal
+     * timerfd is independent of its invocation.
+     */
+    do {
+        nrevents = epoll_pwait(waitsetfd, revents, nevents, -1, NULL);
+    } while (nrevents == -1 && errno == EINTR);
+    if (nrevents > 0) {
+        int orig_nrevents = nrevents;
+        for (int i = 0; i < orig_nrevents; i++)
+            if (revents[i].data.u64 == INTERNAL_TIMERFD)
+                nrevents -= 1;          /* Disregard in total reported events */
+            else
+                ready_set |= (1ULL << revents[i].data.u64);
+    }
+    assert(nrevents >= 0);
+#else /* kqueue */
+    /*
+     * At least one event must be requested in kevent(), otherwise the call
+     * will just return or error.
+     */
+    int nevents = npollfds ? npollfds : 1;
+    int nrevents;
+    uint64_t ready_set = 0;
+    struct kevent revents[nevents];
     struct timespec ts;
-    int rc;
 
     ts.tv_sec = t->timeout_nsecs / 1000000000ULL;
     ts.tv_nsec = t->timeout_nsecs % 1000000000ULL;
 
-    rc = ppoll(pollfds, npollfds, &ts, &pollsigmask);
-    assert(rc >= 0);
-    t->ret = rc;
+    nrevents = kevent(waitsetfd, NULL, 0, revents, nevents, &ts);
+    /*
+     * Unlike the epoll() implementation, we can't easily restart the kqueue()
+     * call on EINTR, due to not having a straightforward way to recalculate
+     * the timeout.  While we could use EVFILT_TIMER similarly to the Linux
+     * timerfd, this has system-wide limits on the number of active timers.
+     *
+     * However: We don't handle any signals, other than by terminating the
+     * tender.  Therefore, we should never see EINTR in practice here. If this
+     * turns out not to be the case, prominently warn the user about it and
+     * pretend we woke up early with no events, which is better than just
+     * asserting/aborting.
+     */
+    if (nrevents == -1 && errno == EINTR) {
+        warnx("hypercall_poll(): kqueue() returned EINTR");
+        warnx("hypercall_poll(): This should not happen, please report a bug");
+        nrevents = 0;
```

Comment:
if this happens repeatedly without delaying, won't the unikernel enter a busy loop at this point?